Ignore files ignored by build

If a file begins with

```
//go:build ignore
// +build ignore
```

it should be ignored by this tool.  Otherwise these comments get stripped, and obviously that causes issues as now a file which should be ignored during build is not.

We end up with an error like

```
#17 0.634 somepackage/somefile.go:7:2: found packages somepackage (fileA.go) and main (fileB.go) in /app
```

--- a/README.md
+++ b/README.md
@@ -82,6 +82,16 @@ func (s Cat) Walk(ctx context.Context) (err error) {
 
 Comments will be stripped. This fits well if your next step is to compile.
 
+### Go compiler directives
+
+Standard Go compiler directives are recognized.
+More details `go help buildconstraint` and https://pkg.go.dev/cmd/go#hdr-Build_constraints.
+
+* `//go:build exclude`
+* `// +build exclude`
+* `//go:build ignore`
+* `// +build ignore`
+
 ### In Development
 
 - [ ] Dynamic error variable name

--- a/internal/example/skipped_buildignore.go
+++ b/internal/example/skipped_buildignore.go
@@ -1,0 +1,12 @@
+//go:build ignore
+// +build ignore
+
+package example
+
+import (
+	"context"
+)
+
+func Skip2(ctx context.Context) (name string, err error) {
+	return "asdf", nil
+}

--- a/internal/example/skipped_gobuildignore.go
+++ b/internal/example/skipped_gobuildignore.go
@@ -1,0 +1,11 @@
+//go:build ignore
+
+package example
+
+import (
+	"context"
+)
+
+func Skip1(ctx context.Context) (name string, err error) {
+	return "asdf", nil
+}

--- a/main.go
+++ b/main.go
@@ -37,6 +37,14 @@ func main() {
 	if err != nil || fileWithComments == nil {
 		log.Fatalf("can not parse input file: %s", err)
 	}
+
+	directives := processor.GoBuildDirectivesFromFile(*fileWithComments)
+	for _, q := range directives {
+		if q.SkipFile() {
+			return
+		}
+	}
+
 	commands, err := processor.CommandsFromFile(*fileWithComments)
 	if err != nil {
 		log.Fatal(err)

--- a/processor/go_build_directive.go
+++ b/processor/go_build_directive.go
@@ -1,0 +1,92 @@
+package processor
+
+import (
+	"go/ast"
+	"strings"
+)
+
+// GoBuildDirective is standard Go compiler directive.
+type GoBuildDirective uint
+
+const (
+	UnknownDirective GoBuildDirective = iota
+	GoBuildIgnore
+	GoBuildExclude
+	BuildIgnore
+	BuildExclude
+)
+
+func (v GoBuildDirective) SkipFile() bool {
+	switch v {
+	case GoBuildIgnore, GoBuildExclude, BuildIgnore, BuildExclude:
+		return true
+	default:
+		return false
+	}
+}
+
+func IsGoBuildIgnore(s string) bool {
+	fs := strings.Fields(s)
+	if len(fs) != 2 {
+		return false
+	}
+	return fs[0] == "//go:build" && fs[1] == "ignore"
+}
+
+func IsGoBuildExclude(s string) bool {
+	fs := strings.Fields(s)
+	if len(fs) != 2 {
+		return false
+	}
+	return fs[0] == "//go:build" && fs[1] == "exclude"
+}
+
+func IsBuildIgnore(s string) bool {
+	fs := strings.Fields(s)
+	if len(fs) != 3 {
+		return false
+	}
+	return fs[0] == "//" && fs[1] == "+build" && fs[2] == "ignore"
+}
+
+func IsBuildExclude(s string) bool {
+	fs := strings.Fields(s)
+	if len(fs) != 3 {
+		return false
+	}
+	return fs[0] == "//" && fs[1] == "+build" && fs[2] == "exclude"
+}
+
+func ParseGoBuildDirective(s string) GoBuildDirective {
+	switch {
+	case IsGoBuildIgnore(s):
+		return GoBuildIgnore
+	case IsGoBuildExclude(s):
+		return GoBuildExclude
+	case IsBuildIgnore(s):
+		return BuildIgnore
+	case IsBuildExclude(s):
+		return BuildExclude
+	default:
+		return UnknownDirective
+	}
+}
+
+func GoBuildDirectivesFromFile(file ast.File) (directives []GoBuildDirective) {
+	for _, q := range file.Comments {
+		if q == nil {
+			continue
+		}
+		for _, c := range q.List {
+			if c == nil {
+				continue
+			}
+			// can not use ast.CommentGroup since its Text() skips directive comments
+			// that do not have empty space after `//` (eg, //go:instrument or //go:generate)
+			if d := ParseGoBuildDirective(c.Text); d != UnknownDirective {
+				directives = append(directives, d)
+			}
+		}
+	}
+	return directives
+}

--- a/processor/go_build_directive_test.go
+++ b/processor/go_build_directive_test.go
@@ -1,0 +1,96 @@
+package processor_test
+
+import (
+	"go/parser"
+	"go/token"
+	"testing"
+
+	"github.com/nikolaydubina/go-instrument/processor"
+)
+
+func TestGoBuildDirectivesFromFile(t *testing.T) {
+	tests := []struct {
+		fileName   string
+		directives []processor.GoBuildDirective
+	}{
+		{
+			fileName: "../internal/example/skipped_buildignore.go",
+			directives: []processor.GoBuildDirective{
+				processor.GoBuildIgnore,
+				processor.BuildIgnore,
+			},
+		},
+		{
+			fileName: "../internal/example/skipped_gobuildignore.go",
+			directives: []processor.GoBuildDirective{
+				processor.GoBuildIgnore,
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.fileName, func(t *testing.T) {
+			fset := token.NewFileSet()
+			f, err := parser.ParseFile(fset, tc.fileName, nil, parser.ParseComments)
+			if err != nil || f == nil {
+				t.Errorf("can not parse input file: %s", err)
+			}
+			directives := processor.GoBuildDirectivesFromFile(*f)
+			if len(directives) != len(tc.directives) {
+				t.Errorf("exp(%#v) != (%#v)", tc.directives, directives)
+			}
+			for i := range directives {
+				if directives[i] != tc.directives[i] {
+					t.Errorf("exp(%#v) != (%#v)", tc.directives, directives)
+				}
+			}
+		})
+	}
+}
+
+func TestParseGoBuildDirective(t *testing.T) {
+	tests := []struct {
+		s string
+		v processor.GoBuildDirective
+	}{
+		{
+			s: "//go:build ignore",
+			v: processor.GoBuildIgnore,
+		},
+		{
+			s: "//go:build exclude",
+			v: processor.GoBuildExclude,
+		},
+		{
+			s: "// +build ignore",
+			v: processor.BuildIgnore,
+		},
+		{
+			s: "// +build exclude",
+			v: processor.BuildExclude,
+		},
+		// whitespace
+		{
+			s: "//go:build    ignore   ",
+			v: processor.GoBuildIgnore,
+		},
+		{
+			s: "//go:build     exclude    ",
+			v: processor.GoBuildExclude,
+		},
+		{
+			s: "//    +build    ignore    ",
+			v: processor.BuildIgnore,
+		},
+		{
+			s: "//    +build    exclude   ",
+			v: processor.BuildExclude,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.s, func(t *testing.T) {
+			if v := processor.ParseGoBuildDirective(tc.s); v != tc.v {
+				t.Errorf("exp(%v) != (%v)", tc.v, v)
+			}
+		})
+	}
+}

